### ISSUE TITLE
Resume training from Lightning generated ckpt alone 

### ISSUE THREAD

**Issue by kotchin** [Lightning-AI/lightning#16014]
### Description & Motivation

It seems that whenever checkpointing is enabled, a corresponding `*.ckpt` file is generated, containing everything from the model weights, to the logging data and the parameters of the LitModule and LitDataModule.

This sounded great to me and so I thought naively that I could simply resume a training running just the following:
`python mainCLI.py fit --ckpt_path /path/to/my/checkpoint.ckpt`

However, this does not work, giving me the following error:
```
usage: mainCLI.py [-h] [-c CONFIG] [--print_config[=flags]] {fit,validate,test,predict,tune} ...
mainCLI.py: error: Configuration check failed :: Key "fit.data.dataset_dir" is required but not included in config object or its value is None.
```
This is an error for a missing path to my dataset, which is contained in the `*.ckpt` file which I'm trying to resume from.

Perhaps this is an intended behavior, or perhaps this is a bug. I'm guessing this has simply not been implemented, but I was wondering if this would be an opportunity to allow for all the parameters to be loaded from the ckpt if it contains them.

There might be possible scenarios which need to be handled though: e.g. if someone provides both a `*.ckpt` file and a config file, which one takes priority? The config file could take priority, allowing for someone to modify some parameters when resuming the training (example expand trainer.max_time if the reason the training ended was due to the trainer have reached maximum time). But if someone can change those parameters, then what parameters are reported by the logger? If someone changes the learning-rate for instance, which learning rate should be logged? I don't have an answer to that, but it seems these questions could also already be asked today, as it's technically already possible to resume a checkpoint differently and modify some parameters. Currently, no change to the parameters are reported, so first parameters are logged and no new parameters are logged.

cc @carmocca @mauvilsa

EDIT: to address the above question: as per @carmocca 's [answer](https://github.com/Lightning-AI/lightning/issues/16014#issuecomment-1348484106) this is not expected to work as thought and outlined above. For this to work, you need to specify the config file using `--config /path/to/my/config.yaml` but this is just a placeholder config file and the content of this config file will in fact be replaced by the parameters contained in the checkpoint.

**Comment by awaelchli** [comment]
Hi @kotchin 
I think the CLI fails to instantiate your LightningDataModule, because it doesn't know what the value for the dataset_dir parameter should be (and it expects it to be in the config file). Would you mind sharing the datamodule implementation, or even better, a small runnable script that shows this error?  

What happens if you edit the config file and add the dataset_dir value manually? Does this get you further?

**Comment by kotchin** [comment]
If I run `python mainCLI.py fit --ckpt_path /path/to/my/checkpoint.ckpt --print_config` it yields the default values for all my parameters instead of the ones actually stored in the ckpt file. 
Indeed it doesn't know what the dataset_dir parameter should be, because I haven't given it any default value. This is why I'm proposing to look for those values in the ckpt file, which was generated by a lightning experiment and contains all values for all parameters for the lightning module and the lightning data module.

If I provide a config file, everything works, which is as expected. But then it ignores all config parameters within the ckpt, which may be what we want, but I was wondering and this is what I'm proposing here, if it would be a good idea to attempt at retrieving all provided parameters from the ckpt which does seem to contain everything that's required for the resuming of the experiment?

**Comment by carmocca** [comment]
One solution would be to use a default `dataset_dir` value and overwrite once the checkpoint gets loaded by implementing `load_state_dict`

**Comment by kotchin** [comment]
@carmocca Thanks for your input:
`overwrite once the checkpoint gets loaded by implementing the load_state_dict()`
Is this done automatically, or does it require a manual intervention?

Digging more into it, it seems a config file is necessary for the LightningCLI to start, but that the end values used are in fact not from the config file but from the checkpoint file, meaning that while all the content of the config file needs to be present to make a hypothetical instantiation of the lightning module and lightning data module, the end result of the instances is following the content of what is defined in the checkpoint, is this understanding correct?

**Comment by carmocca** [comment]
> the end result of the instances is following the content of what is defined in the checkpoint, is this understanding correct?

Yes!

> does it require a manual intervention?

You need to override this hook in your DataModule: https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.core.LightningDataModule.html#pytorch_lightning.core.LightningDataModule.load_state_dict

**Comment by kotchin** [comment]
Thank you for your prompt reply.

In my case, I am working with LightningCLI and I want to reload model weights only and set all other parameters of the LightningModule (and LightningDataModule) with the attributes as defined in the config file, is there a recommended approach for this or is it out of scope for Lightning?
